### DISCUSSION TITLE
Keep getting changed core after RGroupDecompose...

### DISCUSSION THREAD

**Discussion by TAEUKHAM** [rdkit/rdkit#4451] [Q&A] [answered]
Hi! I'm having difficulties about the true functionality of the parameter rdRGroupDecomposition.RGroupDecompositionParameters().onlyMatchAtRGroups.
Would be glad if you help me a little bit:)

Currently, Im trying to rgroup-decompose with r-group specified core.
And don't want to get compound decomposed if the core is different from what I've stated(In R-group manner)

So, I've set
params = rdRGroupDecomposition.RGroupDecompositionParameters()
params.onlyMatchAtRGroups=True
params.removeAllHydrogenRGroups=True
and decomposed but did not get the result I've wanted....
Im keep getting the result of the changed core.

For example, I've setted my core as below.
Core = 'O=P(c1c(Nc2nc(Nc3cc([:3])c([:2])cc3[:1])ncc2[:4])ccc([:5])c1[:6])([:7])[:8]'
And after rgroup-decomposing, I get the succeeded decomposition result with the changed core as below.
Core = 'O=P(c1c(Nc2nc(Nc3cc([:3])c([:2])c([:11])c3[:1])ncc2[:4])c([:10])c([:9])c([:5])c1[:6])([:7])[*:8]'

What should I do in order to prevent the decomposing function of changing my original core?? I really want the function to only decompose with my original core and make fail result if the compound cannot decompose with it.

**Comment by ptosco** [comment]
@TAEUKHAM Can you please provide a reproducible example? You haven’t provided any of the molecules that you have added to the RGroupDecomposition and that resulted in the modified core.
Additionally, your core does not yield a valid molecule:
```
Core = Chem.MolFromSmiles('O=P(c1c(Nc2nc(Nc3cc([:3])c([:2])cc3[:1])ncc2[:4])ccc([:5])c1[:6])([:7])[:8]')
[23:39:14] SMILES Parse Error: syntax error while parsing: O=P(c1c(Nc2nc(Nc3cc([:3])c([:2])cc3[:1])ncc2[:4])ccc([:5])c1[:6])([:7])[:8]
[23:39:14] SMILES Parse Error: Failed parsing SMILES 'O=P(c1c(Nc2nc(Nc3cc([:3])c([:2])cc3[:1])ncc2[:4])ccc([:5])c1[:6])([:7])[:8]' for input: 'O=P(c1c(Nc2nc(Nc3cc([:3])c([:2])cc3[:1])ncc2[:4])ccc([:5])c1[:6])([:7])[:8]'
```

**Comment by TAEUKHAM** [comment]
Sorry for giving you the wrong smile code:(
I've linked my code in markdown format below. Hope you are able to access with it.

In the code, I've stated my core as "[*:7]P([*:8])(=O)c1c([*:9])c([*:10])ccc1Nc1nc(Nc2cc([*:3])c([*:2])cc2[*:1])nc([*:4])c1[*:5]" as smile notation.
However, the decomposition result returned success result with changed core.
I only want the RGroupDecompose function to only decompose with my stated core.

I'm currently changing the parameters of RGroupDecomposition to make this possible but did not make ant effort yet....

[TQB_decompose_test.md](https://github.com/rdkit/rdkit/files/7089157/TQB_decompose_test.md)

**Reply by ptosco** [reply]
@TAEUKHAM 
> In the code, I've stated my core as "[:7]P([:8])(=O)c1c([:9])c([:10])ccc1Nc1nc(Nc2cc([:3])c([:2])cc2[:1])nc([:4])c1[*:5]" as smile notation.

This still fails to yield a valid molecule:
```
Core = Chem.MolFromSmiles("[:7]P([:8])(=O)c1c([:9])c([:10])ccc1Nc1nc(Nc2cc([:3])c([:2])cc2[:1])nc([:4])c1[*:5]")
RDKit ERROR: [09:51:18] SMILES Parse Error: syntax error while parsing: [:7]P([:8])(=O)c1c([:9])c([:10])ccc1Nc1nc(Nc2cc([:3])c([:2])cc2[:1])nc([:4])c1[*:5]
RDKit ERROR: [09:51:18] SMILES Parse Error: Failed parsing SMILES '[:7]P([:8])(=O)c1c([:9])c([:10])ccc1Nc1nc(Nc2cc([:3])c([:2])cc2[:1])nc([:4])c1[*:5]' for input: '[:7]P([:8])(=O)c1c([:9])c([:10])ccc1Nc1nc(Nc2cc([:3])c([:2])cc2[:1])nc([:4])c1[*:5]'
```
Please provide a reproducible example with:
* correct SMILES core
* correct SMILES for a couple of the molecules that you are adding to the R-group decomposition which leads to the modified cores

If you don't, we won't be able to help.

**Reply by TAEUKHAM** [reply]
Im really sorry. I don't know why the asterisk symbol keep disappearing whenever I copy and paste it.
I'll upload my screenshot of my jupyter notebook instead.

![no1](https://user-images.githubusercontent.com/88373558/131638677-af11c845-477a-45af-93e2-e32cdd23f92d.png)
![no2](https://user-images.githubusercontent.com/88373558/131638688-529cd532-d894-4968-a9eb-d4b6566ac588.png)

And this is 10 samples of my embodiments
['COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Br)c(Nc2ccccc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccccc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(F)c(Nc2cccc(F)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccc(Br)cc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2cccc(F)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccc(Cl)cc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(F)c(Nc2cccc(C)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(F)c(Nc2ccc(Br)cc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccc(O)c(F)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Br)c(Nc2ccc(O)c(F)c2P(C)(C)=O)n1']

**Comment by ptosco** [comment, marked as answer]
@TAEUKHAM I cannot reproduce the issue you describe. No additional R groups are added to the core when I use your parameters:
```
from rdkit import Chem
from rdkit.Chem import rdRGroupDecomposition

Core = Chem.MolFromSmiles("[*:7]P([*:8])(=O)c1c([*:9])c([*:10])ccc1Nc1nc(Nc2cc([*:3])c([*:2])cc2[*:1])nc([*:4])c1[*:5]")

mol_smiles = ['COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Br)c(Nc2ccccc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccccc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(F)c(Nc2cccc(F)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccc(Br)cc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2cccc(F)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccc(Cl)cc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(F)c(Nc2cccc(C)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(F)c(Nc2ccc(Br)cc2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Cl)c(Nc2ccc(O)c(F)c2P(C)(C)=O)n1', 'COc1cc(N2CCC(N3CCN(C)CC3)CC2)c(C)cc1Nc1ncc(Br)c(Nc2ccc(O)c(F)c2P(C)(C)=O)n1']

mols = [Chem.MolFromSmiles(smi) for smi in mol_smiles]

params = rdRGroupDecomposition.RGroupDecompositionParameters()
params.onlyMatchAtRGroups = True
params.removeAllHydrogenRGroups = True

rgd = rdRGroupDecomposition.RGroupDecomposition(Core, params)

for mol in mols:
    assert rgd.Add(mol) != -1

rgd.Process()

res = rgd.GetRGroupsAsColumns(asSmiles=True)
set(res["Core"]).pop()
'O=P(c1c(Nc2nc(Nc3cc([*:3])c([*:2])cc3[*:1])nc([*:4])c2[*:5])ccc([*:10])c1[*:9])([*:7])[*:8]'
```

**Reply by TAEUKHAM** [reply]
@ptosco 
Thanks ptosco!! I really appreciate your sincere reply.
I've also checked rdRGroupDecomposition.RGroupDecomposition function with your code and found out the core did not change as you tested but again got wrong result from rdkit.Chem.rdRGroupDecomposition.RGroupDecompose function what I've testing with.

Then, I changed all the H-related parameters to false like the picture below, I've got the result what I've wanted:)!!
I'm not really sure but the "onlyMatchAtRGroups" parameter could be disrupted by the H-related parameters.
Anyway, thanks again for solving my problem:))

![answer github](https://user-images.githubusercontent.com/88373558/131766986-7af27ba0-2b29-4742-81be-23a81170ae19.png)
